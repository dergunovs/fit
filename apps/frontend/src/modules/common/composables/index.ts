import { ref, computed, onMounted } from 'vue';
import { useRouter, useRoute } from 'vue-router';

import LayoutAdmin from '@/common/components/LayoutAdmin.vue';
import LayoutDefault from '@/common/components/LayoutDefault.vue';

export function useLayout() {
  const router = useRouter();
  const route = useRoute();

  const isLoaded = ref(false);

  const layoutComponent = computed(() => (route.meta.layout === 'admin' ? LayoutAdmin : LayoutDefault));

  onMounted(async () => {
    await router.isReady();
    isLoaded.value = true;
  });

  return {
    isLoaded,
    layoutComponent,
  };
}

export function useRouteId(title: string, isQuery?: boolean) {
  const route = useRoute();

  const id = computed(() => {
    const value = isQuery ? route.query[title]?.toString() : route.params[title].toString();

    return value || '';
  });

  return { id };
}
