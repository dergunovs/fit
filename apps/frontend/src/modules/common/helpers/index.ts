import { toast } from 'mhz-ui';

export async function copyToClipboard(text: string, toastText: string) {
  await navigator.clipboard.writeText(text);

  toast.success(toastText);
}
