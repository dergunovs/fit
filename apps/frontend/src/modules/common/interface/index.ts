import { TPostAuthLoginDTO } from 'fitness-tracker-contracts';
import { FunctionalComponent } from 'vue';

export interface INavItem {
  url: string;
  title: string;
  icon: FunctionalComponent;
}

export interface IOnSuccess {
  create?: (id?: string) => Promise<void>;
  update?: () => Promise<void>;
  delete?: () => Promise<void>;
  login?: (user: TPostAuthLoginDTO) => Promise<void>;
  setup?: () => Promise<void>;
  register?: () => Promise<void>;
}

export interface IPromptEvent extends Event {
  prompt: () => Promise<void>;
}

export interface IGetLatestVersionDTO {
  'dist-tags': { latest: string };
}
