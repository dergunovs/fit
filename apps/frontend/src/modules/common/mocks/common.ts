import { vi } from 'vitest';

export const spyCopyToClipboard = vi.spyOn(navigator.clipboard, 'writeText').mockResolvedValue();
