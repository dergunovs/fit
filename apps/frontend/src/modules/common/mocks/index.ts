export * from './helpers';
export * from './ui';
export * from './router';
export * from './layout';
export * from './pwa';
export * from './service';
