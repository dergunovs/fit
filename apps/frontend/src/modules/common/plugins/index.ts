import { createI18n, I18nOptions } from 'vue-i18n';

import ru from '@/common/locales/ru.json';
import en from '@/common/locales/en.json';

const messages = { en, ru } as const;

const options: I18nOptions = { legacy: false, locale: 'ru', messages };

// eslint-disable-next-line @typescript-eslint/no-type-alias
export type TMessageSchema = typeof en;

export const i18n = createI18n<false, typeof options>(options);
