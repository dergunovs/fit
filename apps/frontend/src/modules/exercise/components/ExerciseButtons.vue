<template>
  <div :class="$style.buttons" :data-tall="props.isEdit || props.isPassing">
    <button
      v-if="(props.isEdit || (props.isPassing && !props.isActive && !props.isCurrent)) && props.index !== 0"
      @click="emit('setIndex', props.index)"
      type="button"
      :class="[$style.arrow, $style.pointer, $style.button]"
      :data-tall="props.isEdit || props.isPassing"
      data-test="exercise-buttons-index-down"
    >
      <IconArrow width="16" height="16" :class="$style.rotateArrow" />
    </button>

    <button
      v-if="(props.isEdit || (props.isPassing && !props.isActive)) && !props.isLast"
      @click="emit('setIndex', props.index + 1)"
      type="button"
      :class="[$style.arrow, $style.pointer, $style.button]"
      :data-tall="props.isEdit || props.isPassing"
      data-test="exercise-buttons-index-up"
    >
      <IconArrow width="16" height="16" />
    </button>

    <button
      v-if="props.isEdit && props.isSetCreatable"
      @click="emit('createSet')"
      type="button"
      :class="[$style.button, $style.pointer, $style.success]"
      :data-tall="props.isEdit || props.isPassing"
      data-test="exercise-buttons-create-set"
    >
      +{{ t('set.super') }}
    </button>

    <button
      @click="handleEditRepeats"
      :class="$style.button"
      type="button"
      :data-tall="props.isEdit || props.isPassing"
      data-test="exercise-buttons-repeats"
    >
      x{{ props.repeats }}
    </button>

    <button
      @click="handleEditWeight"
      v-if="props.isEdit ? props.isWeights : !!props.weight"
      :class="$style.button"
      type="button"
      :data-tall="props.isEdit || props.isPassing"
      data-test="exercise-buttons-weight"
    >
      <IconWeight width="14" height="14" /> {{ props.weight }} {{ t('kg') }}
    </button>

    <span
      v-if="!props.isEdit && !props.isDone && !props.isFutureActivity && !props.isPassing"
      :class="[$style.button, $style.error]"
      :data-tall="props.isEdit || props.isPassing"
      data-test="exercise-buttons-not-done"
    >
      <IconFail width="14" height="14" /> {{ t('notDone') }}
    </span>

    <button
      @click="handleToggleIsToFailure"
      v-if="(!props.isEdit && props.isToFailure) || (props.isPassing && props.isActive)"
      type="button"
      :class="[$style.button, $style.isToFailure, props.isToFailure && $style.success]"
      :data-tall="props.isEdit || props.isPassing"
      :data-active="props.isActive"
      data-test="exercise-buttons-to-failure"
    >
      <IconToFailure width="14" height="14" /> {{ t('toFailure') }}
    </button>

    <span
      v-if="!props.isEdit && props.duration"
      :class="$style.button"
      :data-tall="props.isEdit || props.isPassing"
      data-test="exercise-buttons-duration"
    >
      <IconDuration width="14" height="14" /> {{ formatDuration(props.duration, locale) }}
    </span>

    <button
      v-if="props.isEdit || (props.isPassing && !props.isActive && !props.isOnlyOne)"
      @click="emit('delete')"
      type="button"
      :class="[$style.button, $style.pointer, $style.error]"
      :data-tall="props.isEdit || props.isPassing"
      data-test="exercise-buttons-delete"
    >
      {{ t('delete') }}
    </button>
  </div>
</template>

<script setup lang="ts">
import { formatDuration } from 'mhz-helpers';

import IconArrow from '@/common/icons/arrow.svg';
import IconWeight from '@/common/icons/weight.svg';
import IconDuration from '@/common/icons/duration.svg';
import IconToFailure from '@/common/icons/to-failure.svg';
import IconFail from '@/common/icons/fail.svg';

import { useTI18n } from '@/common/composables';

interface IProps {
  repeats: number;
  index: number;
  weight?: number;
  isSetCreatable?: boolean;
  isLast?: boolean;
  isEdit?: boolean;
  isDone?: boolean;
  isPassing?: boolean;
  isToFailure?: boolean;
  duration?: number;
  isFutureActivity?: boolean;
  isWeights?: boolean;
  isActive?: boolean;
  isCurrent?: boolean;
  isOnlyOne?: boolean;
}

interface IEmit {
  delete: [];
  createSet: [];
  setIndex: [index: number];
  setIsToFailure: [isToFailure: boolean];
  editRepeats: [];
  editWeight: [];
}

const props = defineProps<IProps>();
const emit = defineEmits<IEmit>();

const { t, locale } = useTI18n();

function handleEditRepeats() {
  if (props.isEdit || props.isPassing) emit('editRepeats');
}

function handleEditWeight() {
  if (props.isEdit || props.isPassing) emit('editWeight');
}

function handleToggleIsToFailure() {
  if (props.isPassing) emit('setIsToFailure', !props.isToFailure);
}
</script>

<style module lang="scss">
.buttons {
  display: flex;
  align-items: center;
  height: 24px;
  pointer-events: none;
  border: 1px solid var(--color-gray-dark);
  border-radius: 6px;

  &[data-tall='true'] {
    height: 40px;
    pointer-events: all;
  }
}

.button {
  display: flex;
  gap: 2px;
  align-items: center;
  justify-content: center;
  min-width: 32px;
  height: 24px;
  padding: 0 5px;
  font-size: 0.75rem;
  line-height: 1;
  color: var(--color-black);
  cursor: pointer;
  -webkit-user-select: none;
  user-select: none;
  background: none;
  -webkit-tap-highlight-color: transparent;
  -webkit-touch-callout: none;
  border: none;
  border-left: 1px solid var(--color-gray-dark);

  &:first-child {
    border-left: none;
  }

  &:active {
    background: var(--color-gray-light);
    background-color: var(--color-gray-light);
  }

  :focus {
    outline: none;
  }

  &[data-tall='true'] {
    height: 38px;
    padding: 0 6px;
    font-size: 0.875rem;
  }
}

.arrow {
  font-size: 1rem;
  color: var(--color-primary);
}

.rotateArrow {
  rotate: 180deg;
}

.isToFailure {
  color: var(--color-gray-dark);
}

.pointer {
  cursor: pointer;
}

.success {
  color: var(--color-success-dark);
}

.error {
  color: var(--color-error);
}

:global(.dark) {
  .buttons {
    border-color: var(--color-gray-dark-extra);
  }

  .button {
    color: var(--color-white);
    border-left-color: var(--color-gray-dark-extra);
  }

  .isToFailure {
    color: var(--color-gray);
  }

  .success {
    color: var(--color-success);
  }

  .error {
    color: var(--color-error-light);
  }
}
</style>
