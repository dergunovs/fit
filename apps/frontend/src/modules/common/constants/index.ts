import IconExercise from '@/common/icons/nav-exercise.svg?component';
import IconActivity from '@/common/icons/nav-activity.svg?component';
import IconEquipment from '@/common/icons/weight.svg?component';
import IconUser from '@/common/icons/nav-user.svg?component';

import { URL_USER } from '@/user/constants';
import { URL_EXERCISE } from '@/exercise/constants';
import { URL_ACTIVITY_ADMIN } from '@/activity/constants';
import { URL_EQUIPMENT } from '@/equipment/constants';
import { INavItem } from '@/common/interface';
import { URL_SETUP } from '@/auth/constants';

export const NAV_ITEMS: INavItem[] = [
  { _id: '1', url: URL_EXERCISE, title: 'Упражнения', icon: IconExercise },
  { _id: '2', url: URL_ACTIVITY_ADMIN, title: 'Занятия', icon: IconActivity },
  { _id: '3', url: URL_EQUIPMENT, title: 'Оборудование', icon: IconEquipment },
  { _id: '4', url: URL_USER, title: 'Пользователи', icon: IconUser },
];

export const UPDATE_BUTTON_TEXT = 'Обновить';
export const CREATE_BUTTON_TEXT = 'Создать';

export const URL_HOME = '/';
export const URL_ERROR = '/404';

export const GUEST_PAGES = [URL_HOME, URL_SETUP];
