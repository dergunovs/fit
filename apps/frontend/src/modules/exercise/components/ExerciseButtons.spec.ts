import { describe, it, expect, beforeEach, afterEach } from 'vitest';
import { VueWrapper, enableAutoUnmount } from '@vue/test-utils';
import { dataTest, formatDuration } from 'mhz-helpers';

import ExerciseButtons from './ExerciseButtons.vue';

import { wrapperFactory } from '@/common/test';

const indexDown = dataTest('exercise-buttons-index-down');
const indexUp = dataTest('exercise-buttons-index-up');
const createSet = dataTest('exercise-buttons-create-set');
const repeats = dataTest('exercise-buttons-repeats');
const weight = dataTest('exercise-buttons-weight');
const notDone = dataTest('exercise-buttons-not-done');
const toFailure = dataTest('exercise-buttons-to-failure');
const duration = dataTest('exercise-buttons-duration');
const deleteButton = dataTest('exercise-buttons-delete');

const REPEATS = 4;
const INDEX = 2;
const WEIGHT = 16;
const DURATION = 40;

let wrapper: VueWrapper<InstanceType<typeof ExerciseButtons>>;

beforeEach(() => {
  wrapper = wrapperFactory(ExerciseButtons, {
    isEdit: true,
    isSetCreatable: true,
    isWeights: true,
    repeats: REPEATS,
    index: INDEX,
    weight: WEIGHT,
    duration: DURATION,
  });
});

enableAutoUnmount(afterEach);

describe('ExerciseButtons', async () => {
  it('exists', async () => {
    expect(wrapper.findComponent(ExerciseButtons)).toBeTruthy();
  });

  it('matches snapshot', async () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('hides update index buttons if first or last index', async () => {
    expect(wrapper.find(indexDown).exists()).toBe(true);
    expect(wrapper.find(indexUp).exists()).toBe(true);

    await wrapper.setProps({ index: 0 });

    expect(wrapper.find(indexDown).exists()).toBe(false);
    expect(wrapper.find(indexUp).exists()).toBe(true);

    await wrapper.setProps({ index: 1, isLast: true });

    expect(wrapper.find(indexDown).exists()).toBe(true);
    expect(wrapper.find(indexUp).exists()).toBe(false);

    await wrapper.setProps({ index: 0, isLast: true });

    expect(wrapper.find(indexDown).exists()).toBe(false);
    expect(wrapper.find(indexUp).exists()).toBe(false);
  });

  it('hides create set button if set is not creatable', async () => {
    expect(wrapper.find(createSet).exists()).toBe(true);

    await wrapper.setProps({ isSetCreatable: false });

    expect(wrapper.find(createSet).exists()).toBe(false);
  });

  it('hides weight info if no weight', async () => {
    expect(wrapper.find(weight).exists()).toBe(true);

    await wrapper.setProps({ weight: 0, isEdit: false });

    expect(wrapper.find(weight).exists()).toBe(false);
  });

  it('shows repeats and weight', async () => {
    expect(wrapper.find(repeats).text()).toBe(`x${REPEATS}`);
    expect(wrapper.find(weight).text()).toBe(`${WEIGHT} кг`);
  });

  it('hides not done, to failure and duration in edit mode', async () => {
    expect(wrapper.find(notDone).exists()).toBe(false);
    expect(wrapper.find(toFailure).exists()).toBe(false);
    expect(wrapper.find(duration).exists()).toBe(false);
  });

  it('shows not done, to failure and duration in view mode, and hides buttons', async () => {
    await wrapper.setProps({ isEdit: false, isToFailure: true });

    expect(wrapper.find(notDone).exists()).toBe(true);
    expect(wrapper.find(toFailure).exists()).toBe(true);
    expect(wrapper.find(duration).exists()).toBe(true);

    expect(wrapper.find(indexDown).exists()).toBe(false);
    expect(wrapper.find(indexUp).exists()).toBe(false);
    expect(wrapper.find(createSet).exists()).toBe(false);
    expect(wrapper.find(deleteButton).exists()).toBe(false);
  });

  it('hides not done in future activities', async () => {
    expect(wrapper.find(notDone).exists()).toBe(false);

    await wrapper.setProps({ isEdit: false });

    expect(wrapper.find(notDone).exists()).toBe(true);

    await wrapper.setProps({ isEdit: false, isFutureActivity: true });

    expect(wrapper.find(notDone).exists()).toBe(false);
  });

  it('shows formatted duration', async () => {
    await wrapper.setProps({ isEdit: false });

    expect(wrapper.find(duration).text()).toBe(formatDuration(DURATION));
  });

  it('emits updated index', async () => {
    expect(wrapper.emitted()).not.toHaveProperty('setIndex');

    await wrapper.find(indexDown).trigger('click');

    expect(wrapper.emitted('setIndex')).toHaveLength(1);
    expect(wrapper.emitted()['setIndex'][0]).toStrictEqual([INDEX]);

    await wrapper.find(indexUp).trigger('click');

    expect(wrapper.emitted('setIndex')).toHaveLength(2);
    expect(wrapper.emitted()['setIndex'][1]).toStrictEqual([INDEX + 1]);
  });

  it('emites createSet, repeats and weight', async () => {
    expect(wrapper.emitted()).not.toHaveProperty('createSet');
    expect(wrapper.emitted()).not.toHaveProperty('editRepeats');
    expect(wrapper.emitted()).not.toHaveProperty('editWeight');

    await wrapper.find(createSet).trigger('click');

    expect(wrapper.emitted('createSet')).toHaveLength(1);
    expect(wrapper.emitted()['createSet'][0]).toStrictEqual([]);

    await wrapper.find(repeats).trigger('click');

    expect(wrapper.emitted('editRepeats')).toHaveLength(1);
    expect(wrapper.emitted()['editRepeats'][0]).toStrictEqual([]);

    await wrapper.find(weight).trigger('click');

    expect(wrapper.emitted('editWeight')).toHaveLength(1);
    expect(wrapper.emitted()['editWeight'][0]).toStrictEqual([]);
  });

  it('emits delete', async () => {
    expect(wrapper.emitted()).not.toHaveProperty('delete');

    await wrapper.find(deleteButton).trigger('click');

    expect(wrapper.emitted('delete')).toHaveLength(1);
    expect(wrapper.emitted()['delete'][0]).toStrictEqual([]);
  });
});
