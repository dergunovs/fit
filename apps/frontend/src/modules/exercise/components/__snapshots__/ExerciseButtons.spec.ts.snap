// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ExerciseButtons > matches snapshot 1`] = `
"<div class="_buttons_dedec9" data-tall="true"><button type="button" class="_arrow_dedec9 _pointer_dedec9 _button_dedec9" data-edit="true" data-test="exercise-buttons-index-down">
    <anonymous-stub width="16" height="16" class="_rotateArrow_dedec9"></anonymous-stub>
  </button><button type="button" class="_arrow_dedec9 _pointer_dedec9 _button_dedec9" data-edit="true" data-test="exercise-buttons-index-up">
    <anonymous-stub width="16" height="16"></anonymous-stub>
  </button><button type="button" class="_button_dedec9 _pointer_dedec9 _success_dedec9" data-edit="true" data-test="exercise-buttons-create-set"> +Сет</button><button class="_button_dedec9" type="button" data-edit="true" data-test="exercise-buttons-repeats"> x4</button><button class="_button_dedec9" type="button" data-edit="true" data-test="exercise-buttons-weight">
    <anonymous-stub width="14" height="14"></anonymous-stub> 16 кг
  </button>
  <!--v-if-->
  <!--v-if-->
  <!--v-if--><button type="button" class="_button_dedec9 _pointer_dedec9 _error_dedec9" data-edit="true" data-test="exercise-buttons-delete">Удалить</button>
</div>"
`;
